Add common centroid capacitor test cases

stevenmburns opened the pull request.

stevenmburns commented:
@soneryaldiz @parijatm I figured out how to enable `googletest`-based unit tests.
You just need to build with:
```
pip install -e .[test] --no-build-isolation --install-option='-DBUILD_TESTING=ON'
```
The tests weren't building before because we deleted some functions referenced in the tests.
They are working now. To run the unit tests, try:
```
tests/test_PnR
```
I started to add a test just for the `KeepOrdering` method.
It is in `PlaceRouteHierFlow/placer/unit_tests.cpp`.

stevenmburns commented:
@soneryaldiz @854768750 Here is a unit test the has only four cells but still causes the same kind of infinite loop.
This test is a simplified version for the failing test (infinite loop) in #1026 
See instructions above on how to run it.

stevenmburns commented:
This is the test:
```
TEST(SeqPairTest, KeepOrdering) {
    spdlog::set_level(spdlog::level::debug);
    SeqPair sp;

    std::vector<int> posPair = {0,1,2,3};
    std::vector<int> negPair = {0,1,2,3};

    sp.posPair = posPair;
    sp.negPair = negPair;

    sp.PrintSeqPair();

    design d;

    std::vector<int> counterparts = {1,0,-1,-1};

    for (const auto& counterpart : counterparts) {
      design::block b;
      b.counterpart = counterpart;
      std::vector<design::block> blocks;
      blocks.push_back(b);
      d.Blocks.push_back(blocks);
    }

    std::vector<std::pair<int,int> > v_orderings =  {
      {0, 3},
      {1, 3},
      {3, 2}
    };

    std::vector<std::pair<int,int> > h_orderings =  {
      {0, 1}
    };

    for (const auto& ordering : v_orderings) {
      d.Ordering_Constraints.push_back(make_pair(ordering, placerDB::V));
    }
    for (const auto& ordering : h_orderings) {
      d.Ordering_Constraints.push_back(make_pair(ordering, placerDB::H));
    }

    sp.KeepOrdering(d);


};
```
I'm not really sure what we are trying to do here and what the counterparts are. @854768750 Can you explain that to us? What is the purpose of the `KeepOrdering` method?

srini229 commented:
@stevenmburns My understanding is: `KeepOrdering` method is used to enforce the `ordering` constraint in the sequence pair. When a perturbation of sequence pairs violates ordering, this routine permutes the indices to enforce ordering.

This can be done constructively, since we only perturb one or two indices in a sequence pair at a time.
For example when blocks `A` has to be above `B`, then the index of `A` in: positive seq `< index(B)` and negative seq `> index(B)`. This can easily be enforced during the perturbation `KeepOrdering` routine can be removed.


When two blocks are part of a symmetric pair, each one is other's `counterpart`.

854768750 commented:
For `KeepOrderingShort`, as we have v_ordering
```
{0, 2},
      {2, 1}
```,
0 and 1 cannot be counterpart.
It would be `counterparts = {-1,-1,-1}`

stevenmburns commented:
@854768750 We can't have infinite loops even if the constraints are wrong.
What is your argument that the `do { } while ()` loop in KeepOrdering is going to terminate?
My example is just a pruned down version of Soner's testcase on the other PR.
You can uses this unit test facility to try other cases.

854768750 commented:
It terminates when all constraints are satisfied.

soneryaldiz commented:
A constructive approach sounds more promising and could possibly detect infeasible scenarios. @854768750 , are you planning to prototype this for #1026 ?

stevenmburns commented:
I had a failure of test_cmp_slow in CI. It runs locally but it is slow. Do we want to remove this from the CI?
```
----------------------------------------- generated xml file: /home/smburns/DARPA/ALIGN-public/junit.xml -----------------------------------------
============================================================== slowest 10 durations ==============================================================
272.58s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_slow
36.89s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_fast
24.36s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_3
1.53s teardown tests/pdk/finfet_pdk/test_placer.py::test_cmp_slow
0.93s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_1
0.88s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_2
0.07s teardown tests/pdk/finfet_pdk/test_placer.py::test_hang_3
0.00s teardown tests/pdk/finfet_pdk/test_placer.py::test_hang_2
0.00s setup    tests/pdk/finfet_pdk/test_placer.py::test_cmp_fast
0.00s setup    tests/pdk/finfet_pdk/test_placer.py::test_place_cmp_2
```

soneryaldiz commented:
Let me take a quick look.

soneryaldiz commented:
@stevenmburns , I prefer to keep these tests for now as this has been a problematic area and the related code will likely to be refactored soon. 

On master branch with `pip install -v --no-build-isolation -e . --no-deps --install-option='--build-type=Release`:

It takes:
```
=== slowest 5 durations ===
34.71s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_slow
25.23s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_fast
9.33s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_3
1.02s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_1
0.99s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_2
```

When I reduce log to info, it takes less:
```
=== slowest 5 durations ===
24.59s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_fast
24.07s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_slow
7.98s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_3
1.42s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_1
1.05s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_2
```

When I stop flow before route,  it takes less:
```
=== slowest 5 durations ===
24.49s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_fast
24.13s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_slow
8.40s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_3
1.14s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_1
1.12s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_2
```

I will commit the changes to this PR shortly.

soneryaldiz commented:
Instead of skipping the test, could we do a fast build for checkin instead of doing a developer build?

stevenmburns commented:
@soneryaldiz Why do we every have the log level set to debug for CI?

stevenmburns commented:
@soneryaldiz It never occurred to me that debug level would be set in a CI run. The reason for the long run time was the debug level stuff I had added to KeepOrdering. I removed those.

stevenmburns commented:
This is what we have now for durations:
```
----------------------------------------- generated xml file: /home/smburns/DARPA/ALIGN-public/junit.xml -----------------------------------------
============================================================== slowest 10 durations ==============================================================
17.66s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_fast
17.11s call     tests/pdk/finfet_pdk/test_placer.py::test_cmp_slow
5.81s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_3
0.90s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_1
0.67s call     tests/pdk/finfet_pdk/test_placer.py::test_hang_2
0.00s setup    tests/pdk/finfet_pdk/test_placer.py::test_cmp_slow
0.00s setup    tests/pdk/finfet_pdk/test_placer.py::test_cmp_fast
0.00s setup    tests/pdk/finfet_pdk/test_placer.py::test_hang_1
0.00s setup    tests/pdk/finfet_pdk/test_placer.py::test_hang_2
0.00s setup    tests/pdk/finfet_pdk/test_placer.py::test_hang_3

Results (25.69s):
       5 passed
      13 skipped
```

soneryaldiz commented:
What is probably further slowing things down is that parallel threads are writing to the same log file: #755

stevenmburns commented:
@soneryaldiz This should be ready to merge.

stevenmburns commented:
@soneryaldiz @parijatm @854768750 @Lastdayends @srini229 To use the resurrected C++ unit test facility, we need to compile using the following command sequence:
```
pip install -e .[test] --no-build-isolation
pip install -v --no-build-isolation -e . --no-deps --install-option='--build-type=Release' --install-option='-DBUILD_TESTING=ON'
```
(You can replace `Release` with `Debug` or `RelWithDebInfo`. This is also in README.md now.)
Then the unit tests are collected and put into an executable in `tests/test_PnR`.
You can just run this, but pytest also discovers it and runs it when you run `pytest` in the ALIGN-public directory.
```

854768750 commented:
Yes. This test case has a valid solution. It may need another seqpair ordering method to solve this.




------------------&nbsp;Original&nbsp;------------------
From: "Steve ***@***.***&gt;; 
Date: 2022年3月18日(星期五) 上午9:17
To: ***@***.***&gt;; 
Cc: ***@***.***&gt;; ***@***.***&gt;; 
Subject: Re: [ALIGN-analoglayout/ALIGN-public] Enable C++ Unit Testing and Coverage (PR #1027)





 
@854768750 I pretty sure the following test case (currently disabled) represents a valid configuration:
 TEST(SeqPairTest, DISABLED_KeepOrdering) {     auto save_level = spdlog::get_level();     spdlog::set_level(spdlog::level::trace);     SeqPair sp;     std::vector<int&gt; posPair = {0,1,2,3};     std::vector<int&gt; negPair = {0,1,2,3};     sp.posPair = posPair;     sp.negPair = negPair;     sp.PrintSeqPair();     design d;     /*      *           0   |   1      *               3      *               2      */     std::vector<int&gt; counterparts = {1,0,-1,-1};          for (const auto&amp; counterpart : counterparts) {       design::block b;       b.counterpart = counterpart;       std::vector<design::block&gt; blocks;       blocks.push_back(b);       d.Blocks.push_back(blocks);     }     std::vector<std::pair<int,int&gt; &gt; v_orderings =  {       {0, 3},       {1, 3},       {3, 2}     };     std::vector<std::pair<int,int&gt; &gt; h_orderings =  {       {0, 1}     };     for (const auto&amp; ordering : v_orderings) {       d.Ordering_Constraints.push_back(make_pair(ordering, placerDB::V));     }     for (const auto&amp; ordering : h_orderings) {       d.Ordering_Constraints.push_back(make_pair(ordering, placerDB::H));     }     sp.KeepOrdering(d);     spdlog::set_level(save_level); };  
The block 0 and 1 have a Symmetric block constraint and a horizontal ordering constraint.
 There are vertical ordering constraints between 0,1 and 3, and between 3 and 2.
 Depending on the starting permuation for negPir, the current algorithm will go into an infinite loop.
 You can check this by removing the "DISABLED_" prefix from the test name.
 
We are going to check this PR in. You would open a new PR to work on fixing this error.
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android. 
You are receiving this because you were mentioned.Message ID: ***@***.***&gt;

stevenmburns commented:
It seems that you can run currently disabled tests using the following command:
```
tests/test_PnR --gtest_also_run_disabled_tests
```

soneryaldiz reviewed (approved):
Thanks for enabling and demonstrating unit tests. The conflict is resolved.
